### ISSUE TITLE
Super new idea: the file voucher

### ISSUE THREAD

**Issue by DiegoPino** [esmero/strawberryfield#76]
# What is this?

Late night idea i came up with! So, imagine you have a 1TB file you need to attach to an ADO. Imagine that 1TB file is a WARC, or a huge movie, who knows, all your childhood love letters in 600dpi TIFFS and PDFS zipped. Who cares. Well the owner cares. OK, the fact is those are large files. You for sure don't want to upload that via the UI. But you can use the multipart upload system of S3, drag and drop it directly into min.io or even FTP it. Let's say you uploaded it. Now how in the world will you attach it to your to be born ADO when filling up a webform? Drupal only allows you to upload files. And selecting files from Drupal (when you have a million...) is not  such a good idea, like via a Autocomplete..(but could be done..), but there is this idea...

## The File Voucher makes its debut! 

Idea is: when you drop such a large file into S3, S3 tells Drupal, hey friend, there is a new file, and Drupal generates, a one time, `voucher` for you. `Voucher` is of course digital, a tiny unique hash or number. We can setup different folders per user or role in S3 and depending on where things get dropped, via a `webhook`, Drupal gets notified and Drupal notifies the user (via a Drupal message and an alert/view in their user profile). Now when creating the new ADO, the user instead of uploading a file, or pointing to a hard to remember URI (maybe the user does not even know the URI right?) just adds that voucher (can select from his available ones or paste one someone send him to a special field). And all the rest, attaching, classifying, JSON-ifying etc happens magic-ly. The Voucher expires and now the file becomes a fully Drupal driven entity.

I know i make this sound like incredible when its just so simple stuff. But guess what. Nobody has this! And its just super useful. Admins can upload large files at night and then pass vouchers to Metadata people. Vouchers can be resolved before being used so the user knows what the thing is.

Saturday and late. But i had this idea a few days ago and needed to bring it somewhere. Hope we get this done, sounds like a good use case!

@giancarlobi @marlo-longley maybe this makes sense to you

**Comment by dmer** [comment]
found my way here via https://github.com/esmero/strawberryfield/issues/107 
This is a not just a good idea, but might be a functional requirement for some folks.
I was just looking at the S3 multipart upload b/c we're going to have to use that for some of the files in the CAR project and it would be very nice to have an easy-to-use connection to that file for the form.

As I'm writing this there's another thought that is not well-formed, but I'll try - I like the notion that you might be able to choose the voucher from a list of available ones - I'm picturing a possible workflow with a "pending" queue that a user can grab tokens.

Also wondering if this could work w/ AMI? For example could I load all my files to S3, get tokens and then feed them in w/ a spreadsheet?

**Comment by DiegoPino** [comment]
Yes! I'm actually on that code planning right now and may need more feedback. I had some open ended ideas but happy to hear more workflows and expectations (the simpler for the user the better)

**Comment by DiegoPino** [comment]
@pcambra I will assign this macro ISSUE to you. It will also entail at least the webform_strawberryfield Module too. As I explained via email it entails more and is a project in itself. I estimate more than 30 days for the whole thing but we can move in steps. You can copy your comments/ideas regarding my email here too. I will also add its own Milestone as a sub-project so we can track this between releases/versions. Thanks a lot for taking this one

**Comment by pcambra** [comment]
This is my free translation with my own comments of the thoughts that @DiegoPino has shared with me via email and slack:

**Bottomline as is mentioned here is to be able to upload humongous files before ingesting the content via the webform widget**

## Upload

This not only will serve the needs of creating the ADOs via the Drupal interface (Node/Webform) but also through API. This task can be broken into several parts:

* Upload via the Drupal UI: We need upload by chunks. plupload is an option but there are other solutions based in JS. Comments from Diego include that the S3 dependency can be assumed and this needs to allow [optional] validation of the checksum at the end of the upload.
* API upload. We need an endpoint that requires authentication and allows bulk/chunk uploads.
* Direct upload to storage (complexity ++):
  * Leveraging Minio for S3 so we can use events. Depending on where the file was uploaded and its properties, a new voucher is generated. How do we notify the uploader? email? do we create the user?
  * S3 direct would allow us to use SNS and tunnel this from archipelago.
  * Cron based. This would be basically "brute forcing" by checking every X minutes/hours certain folders (user based?), sensible to date.

Personal note: maybe if we're unsure of what approach to take, we could leave the door open and implement generic processing that then plugins can implement depending on the site options. Another thought I'm having is that maybe the voucher doesn NOT come first to the ADO creation but it is the ADO creation the one specifying "I'm going to use a voucher", then the editor selects *how* this voucher is going to be generated. Finally, through a simplified edit process, the link is made between ADO and file. This would simplify a lot the process, specially the cron based one.

-----

Some preliminary implementation details:

- [Must] Voucher custom entity that implements the logic of the voucher (maybe this could be a handler that can be altered/chosen), forms and permissions
- [Must] A custom widget for uploading the files. Needs to support multipart and resume?
- [Must] Webhook controllers (different depending on the implementation technique, SNS, minio, etc) so notifications are processed.
- [Open question] AMI integration through a service. I would need more information about this.
- [Maybe] Voucher webform element that links the file with the ADO. Maybe it could be an action instead?
- [Misc] Various forms and config

**Comment by DiegoPino** [comment]
@pcambra I like your approach here. The question of before or at the end will probably have the most complex answer: both. I think this is a good @pcambra + @alliomeria collaboration to find out the different CRUD (create/edit/update/delete) workflows and how/when a Voucher can integrate via the UI/messages and User interaction. This also speaks about maybe different roles users can have in a repo and having a more generalized (as you correctly said) api/services around this may even allow in the future to think of "external systems" adding Vouchers others than the ones we have described here:

My guts tell me we should start with

- [x] Voucher custom entity that implements the logic of the voucher (maybe this could be a handler that can be altered/chosen), forms and permissions. I would add here also: the storage of course and also "the life cycle" as a inherent thing since once used they should In time expire (maybe? Allison may know more.. maybe not immediately thinking of the Batch ingest use case in AMI/Spreadsheet where the same Batch could run a few times while changes/fixes are being made)

What do you think? Does this, more invisible to UI/UX part sounds good? Please feel free to open an ISSUE for that, you get all my help/support/code on this. HUGS!
